update to 0.1.46 break my codes

my project https://github.com/Sherlock-Holo/rfs/tree/new-fuse use your `async-trait` to create a trait with async fn, but today some problems happened.

When I run `cargo update` it updates the `async-trait` version to 0.1.46 and `cargo clippy` reports errors

> ```
> warning: the feature `generic_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
>  --> src/lib.rs:2:12
>   |
> 2 | #![feature(generic_associated_types)]
>   |            ^^^^^^^^^^^^^^^^^^^^^^^^
>   |
>   = note: `#[warn(incomplete_features)]` on by default
>   = note: see issue #44265 <https://github.com/rust-lang/rust/issues/44265> for more information
> 
> error[E0308]: mismatched types
>     --> src/client/filesystem.rs:2004:26
>      |
> 229  |       type DirEntryPlusStream = impl Stream<Item = Result<DirectoryEntryPlus>> + Send;
>      |                                 ----------------------------------------------------- the expected opaque type
> ...
> 1971 |               let entries = dir_entries.into_iter().filter_map(move |dir_entry| {
>      |  ______________________________________________________________-
> 1972 | |                 let attr = if let Some(attr) = dir_entry.attr {
> 1973 | |                     attr
> 1974 | |                 } else {
> ...    |
> 2000 | |                 })
> 2001 | |             });
>      | |_____________- the found closure
> ...
> 2004 |                   entries: stream::iter(entries).map(Ok),
>      |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected opaque type, found struct `futures_util::stream::Map`
>      |
>      = note: expected opaque type `impl std::marker::Send+tokio_stream::Stream`
>                      found struct `futures_util::stream::Map<futures_util::stream::Iter<std::iter::FilterMap<std::vec::IntoIter<pb::read_dir_response::DirEntry>, [closure@src/client/filesystem.rs:1971:62: 2001:14]>>, fn(fuse3::path::reply::DirectoryEntryPlus) -> std::result::Result<fuse3::path::reply::DirectoryEntryPlus, _> {std::result::Result::<fuse3::path::reply::DirectoryEntryPlus, _>::Ok}>`
> 
> error[E0698]: type inside `async` block must be known in this context
>     --> src/client/filesystem.rs:1916:18
>      |
> 1916 |         for _ in 0..3 {
>      |                  ^ cannot infer type for type `{integer}`
>      |
> note: the type is part of the `async` block because of this `await`
>     --> src/client/filesystem.rs:1948:25
>      |
> 1948 |                         time::sleep(Duration::from_secs(1)).await;
>      |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error[E0698]: type inside `async` block must be known in this context
>     --> src/client/filesystem.rs:1916:21
>      |
> 1916 |         for _ in 0..3 {
>      |                     ^ cannot infer type for type `{integer}`
>      |
> note: the type is part of the `async` block because of this `await`
>     --> src/client/filesystem.rs:1948:25
>      |
> 1948 |                         time::sleep(Duration::from_secs(1)).await;
>      |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error[E0698]: type inside `async` block must be known in this context
>     --> src/client/filesystem.rs:1916:18
>      |
> 1916 |         for _ in 0..3 {
>      |                  ^^^^ cannot infer type for type `{integer}`
>      |
> note: the type is part of the `async` block because of this `await`
>     --> src/client/filesystem.rs:1948:25
>      |
> 1948 |                         time::sleep(Duration::from_secs(1)).await;
>      |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error[E0698]: type inside `async` block must be known in this context
>     --> src/client/filesystem.rs:1916:18
>      |
> 1916 |         for _ in 0..3 {
>      |                  ^^^^ cannot infer type for type `{integer}`
>      |
> note: the type is part of the `async` block because of this `await`
>     --> src/client/filesystem.rs:1948:25
>      |
> 1948 |                         time::sleep(Duration::from_secs(1)).await;
>      |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error: could not find defining uses
>    --> src/client/filesystem.rs:229:31
>     |
> 229 |     type DirEntryPlusStream = impl Stream<Item = Result<DirectoryEntryPlus>> + Send;
>     |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error: aborting due to 6 previous errors; 1 warning emitted
> 
> Some errors have detailed explanations: E0308, E0698.
> For more information about an error, try `rustc --explain E0308`.
> error: could not compile `rfs`
> 
> To learn more, run the command again with --verbose.
> ```

I think the important point is the `mismatched types`

When I downgrade `async-trait` to 0.1.42, `cargo clippy` doesn't report errors.

> ```
>    Compiling rfs v0.4.0 (/home/sherlock/git/rfs)
> warning: the feature `generic_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
>  --> src/lib.rs:2:12
>   |
> 2 | #![feature(generic_associated_types)]
>   |            ^^^^^^^^^^^^^^^^^^^^^^^^
>   |
>   = note: `#[warn(incomplete_features)]` on by default
>   = note: see issue #44265 <https://github.com/rust-lang/rust/issues/44265> for more information
> 
> warning: this function has too many arguments (9/7)
>     --> src/client/filesystem.rs:1447:5
>      |
> 1447 | /     async fn getlk(
> 1448 | |         &self,
> 1449 | |         _req: Request,
> 1450 | |         _path: Option<&OsStr>,
> ...    |
> 1456 | |         pid: u32,
> 1457 | |     ) -> Result<ReplyLock> {
>      | |__________________________^
>      |
>      = note: `#[warn(clippy::too_many_arguments)]` on by default
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments
> 
> warning: this function has too many arguments (10/7)
>     --> src/client/filesystem.rs:1542:5
>      |
> 1542 | /     async fn setlk(
> 1543 | |         &self,
> 1544 | |         req: Request,
> 1545 | |         _path: Option<&OsStr>,
> ...    |
> 1552 | |         block: bool,
> 1553 | |     ) -> Result<()> {
>      | |___________________^
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments
> 
> warning: this function has too many arguments (10/7)
>     --> src/client/filesystem.rs:2033:5
>      |
> 2033 | /     async fn copy_file_range(
> 2034 | |         &self,
> 2035 | |         _req: Request,
> 2036 | |         _from_path: Option<&OsStr>,
> ...    |
> 2043 | |         flags: u64,
> 2044 | |     ) -> Result<ReplyCopyFileRange> {
>      | |___________________________________^
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments
> 
> warning: manual implementation of `Option::map`
>    --> src/server/rpc/server.rs:774:20
>     |
> 774 |               atime: if let Some(atime) = attr.access_time {
>     |  ____________________^
> 775 | |                 Some(convert_proto_time_to_system_time(Some(atime)))
> 776 | |             } else {
> 777 | |                 None
> 778 | |             },
>     | |_____________^ help: try this: `attr.access_time.map(|atime| convert_proto_time_to_system_time(Some(atime)))`
>     |
>     = note: `#[warn(clippy::manual_map)]` on by default
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
> 
> warning: manual implementation of `Option::map`
>    --> src/server/rpc/server.rs:779:20
>     |
> 779 |               mtime: if let Some(mtime) = attr.modify_time {
>     |  ____________________^
> 780 | |                 Some(convert_proto_time_to_system_time(Some(mtime)))
> 781 | |             } else {
> 782 | |                 None
> 783 | |             },
>     | |_____________^ help: try this: `attr.modify_time.map(|mtime| convert_proto_time_to_system_time(Some(mtime)))`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
> 
> warning: manual implementation of `Option::map`
>    --> src/server/rpc/server.rs:784:20
>     |
> 784 |               ctime: if let Some(ctime) = attr.change_time {
>     |  ____________________^
> 785 | |                 Some(convert_proto_time_to_system_time(Some(ctime)))
> 786 | |             } else {
> 787 | |                 None
> 788 | |             },
>     | |_____________^ help: try this: `attr.change_time.map(|ctime| convert_proto_time_to_system_time(Some(ctime)))`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
> 
> warning: 7 warnings emitted
> 
>     Finished dev [unoptimized + debuginfo] target(s) in 1.77s
> ```

I try to write the smallest reproduction codes but failed :(

Could you tell me how to fix this problem?

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -3,12 +3,13 @@ use crate::parse::Item;
 use crate::receiver::{has_self_in_block, has_self_in_sig, mut_pat, ReplaceSelf};
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote, quote_spanned, ToTokens};
+use std::collections::BTreeSet as Set;
 use syn::punctuated::Punctuated;
-use syn::visit_mut::VisitMut;
+use syn::visit_mut::{self, VisitMut};
 use syn::{
     parse_quote, Attribute, Block, FnArg, GenericParam, Generics, Ident, ImplItem, Lifetime, Pat,
     PatIdent, Receiver, ReturnType, Signature, Stmt, Token, TraitItem, Type, TypeParamBound,
-    WhereClause,
+    TypePath, WhereClause,
 };
 
 macro_rules! parse_quote_spanned {
@@ -34,6 +35,7 @@ enum Context<'a> {
     },
     Impl {
         impl_generics: &'a Generics,
+        associated_type_impl_traits: &'a Set<Ident>,
     },
 }
 
@@ -71,7 +73,7 @@ pub fn expand(input: &mut Item, is_local: bool) {
                         method.attrs.push(parse_quote!(#[must_use]));
                         if let Some(block) = block {
                             has_self |= has_self_in_block(block);
-                            transform_block(sig, block);
+                            transform_block(context, sig, block);
                             method.attrs.push(lint_suppress_with_body());
                         } else {
                             method.attrs.push(lint_suppress_without_body());
@@ -90,8 +92,18 @@ pub fn expand(input: &mut Item, is_local: bool) {
             let elided = lifetimes.elided;
             input.generics.params = parse_quote!(#(#elided,)* #params);
 
+            let mut associated_type_impl_traits = Set::new();
+            for inner in &input.items {
+                if let ImplItem::Type(assoc) = inner {
+                    if let Type::ImplTrait(_) = assoc.ty {
+                        associated_type_impl_traits.insert(assoc.ident.clone());
+                    }
+                }
+            }
+
             let context = Context::Impl {
                 impl_generics: &input.generics,
+                associated_type_impl_traits: &associated_type_impl_traits,
             };
             for inner in &mut input.items {
                 if let ImplItem::Method(method) = inner {
@@ -99,7 +111,7 @@ pub fn expand(input: &mut Item, is_local: bool) {
                     if sig.asyncness.is_some() {
                         let block = &mut method.block;
                         let has_self = has_self_in_sig(sig) || has_self_in_block(block);
-                        transform_block(sig, block);
+                        transform_block(context, sig, block);
                         transform_sig(context, sig, has_self, false, is_local);
                         method.attrs.push(lint_suppress_with_body());
                     }
@@ -296,7 +308,7 @@ fn transform_sig(
 //
 //         ___ret
 //     })
-fn transform_block(sig: &mut Signature, block: &mut Block) {
+fn transform_block(context: Context, sig: &mut Signature, block: &mut Block) {
     if let Some(Stmt::Item(syn::Item::Verbatim(item))) = block.stmts.first() {
         if block.stmts.len() == 1 && item.to_string() == ";" {
             return;
@@ -345,18 +357,24 @@ fn transform_block(sig: &mut Signature, block: &mut Block) {
     }
 
     let stmts = &block.stmts;
-    let let_ret = match &sig.output {
+    let let_ret = match &mut sig.output {
         ReturnType::Default => quote_spanned! {block.brace_token.span=>
             let _: () = { #(#decls)* #(#stmts)* };
         },
-        ReturnType::Type(_, ret) => quote_spanned! {block.brace_token.span=>
-            if let ::core::option::Option::Some(__ret) = ::core::option::Option::None::<#ret> {
-                return __ret;
+        ReturnType::Type(_, ret) => {
+            if contains_associated_type_impl_trait(context, ret) {
+                quote!(#(#decls)* #(#stmts)*)
+            } else {
+                quote_spanned! {block.brace_token.span=>
+                    if let ::core::option::Option::Some(__ret) = ::core::option::Option::None::<#ret> {
+                        return __ret;
+                    }
+                    let __ret: #ret = { #(#decls)* #(#stmts)* };
+                    #[allow(unreachable_code)]
+                    __ret
+                }
             }
-            let __ret: #ret = { #(#decls)* #(#stmts)* };
-            #[allow(unreachable_code)]
-            __ret
-        },
+        }
     };
     let box_pin = quote_spanned!(block.brace_token.span=>
         Box::pin(async move { #let_ret })
@@ -378,6 +396,41 @@ fn has_bound(supertraits: &Supertraits, marker: &Ident) -> bool {
         }
     }
     false
+}
+
+fn contains_associated_type_impl_trait(context: Context, ret: &mut Type) -> bool {
+    struct AssociatedTypeImplTraits<'a> {
+        set: &'a Set<Ident>,
+        contains: bool,
+    }
+
+    impl<'a> VisitMut for AssociatedTypeImplTraits<'a> {
+        fn visit_type_path_mut(&mut self, ty: &mut TypePath) {
+            if ty.qself.is_none()
+                && ty.path.segments.len() == 2
+                && ty.path.segments[0].ident == "Self"
+                && self.set.contains(&ty.path.segments[1].ident)
+            {
+                self.contains = true;
+            }
+            visit_mut::visit_type_path_mut(self, ty);
+        }
+    }
+
+    match context {
+        Context::Trait { .. } => false,
+        Context::Impl {
+            associated_type_impl_traits,
+            ..
+        } => {
+            let mut visit = AssociatedTypeImplTraits {
+                set: associated_type_impl_traits,
+                contains: false,
+            };
+            visit.visit_type_mut(ret);
+            visit.contains
+        }
+    }
 }
 
 fn where_clause_or_default(clause: &mut Option<WhereClause>) -> &mut WhereClause {

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -1,6 +1,6 @@
 #![cfg_attr(
     async_trait_nightly_testing,
-    feature(min_specialization, min_const_generics)
+    feature(min_specialization, min_const_generics, type_alias_impl_trait)
 )]
 #![allow(
     clippy::let_underscore_drop,
@@ -1276,5 +1276,27 @@ pub mod issue149 {
         async fn fail() -> &'static dyn Ret {
             return &Thing;
         }
+    }
+}
+
+// https://github.com/dtolnay/async-trait/issues/152
+#[cfg(async_trait_nightly_testing)]
+pub mod issue152 {
+    use async_trait::async_trait;
+
+    #[async_trait]
+    trait Trait {
+        type Assoc;
+
+        async fn f(&self) -> Self::Assoc;
+    }
+
+    struct Struct;
+
+    #[async_trait]
+    impl Trait for Struct {
+        type Assoc = impl Sized;
+
+        async fn f(&self) -> Self::Assoc {}
     }
 }

